### PR TITLE
[RDY] version.c: update + upstream patch porting

### PR DESCRIPTION
**Pull request by brcolow**
Port upstream vim patches 955, 974, 975, and 989. Mark patches 964, 968, 970, 971 and 984 as NA.

patch 7.4.955
Problem:    Vim doesn't recognize .pl6 and .pod6 files.
Solution:   Recognize them as perl6 and pod6. (Mike Eve)

patch 7.4.974
Problem:    When using :diffsplit the cursor jumps to the first line.
Solution:   Put the cursor on the line related to where the cursor was before
            the split.

patch 7.4.975
Problem:    Using ":sort" on a very big file sometimes causes text to be
            corrupted. (John Beckett)
Solution:   Copy the line into a buffer before calling ml_append().

patch 7.4.989
Problem:    Leaking memory when hash_add() fails. Coverity error 99126.
Solution:   When hash_add() fails free the memory.

778 marked as not NA as it will be needed once vim patch 754 is merged
968 marked as NA because it is related to a deleted tests (test86/test87)
970 marked as NA because it is guarded by: `# if defined(FEAT_GUI_GTK) || defined(PROTO`
    and is inside a function that no longer exists
971 marked as NA because it relates to sorting the function table, which
    is already correctly sorted

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -7097,6 +7097,7 @@ static struct fst {
   {"arglistid",       0, 2, f_arglistid},
   {"argv",            0, 1, f_argv},
   {"asin",            1, 1, f_asin},    /* WJMc */
+  {"asin",            1, 1, f_asin},    /* WJMc */
```
Comment (justinmk) [inline src/nvim/eval.c]: Why repeat this?

**Comment by justinmk** [comment]
You are right about 968, test86/test87 were deleted in 0e65caa40ed9dbd1e06d023cd7eadce16f444f30.

> 970 because it is guarded by:
> # if defined(FEAT_GUI_GTK) || defined(PROTO.

:+1: Also right, and `init_preedit_start_col` doesn't exist.

**Comment by brcolow** [comment]
I was clearly going too fast =). It looks like the patch is simply about *sorting* the function table, not adding a function. Since it is already sorted correctly in Neovim, I marked it as NA. So this PR instead marks 3 patches as NA.

**Comment by justinmk** [comment]
Ok, might as well keep this around for a couple days to collect more NAs.

**Comment by Shougo** [comment]
@brcolow :+1:

**Comment by wsdjeg** [comment]
@brcolow @Shougo @justinmk 
happy new year!

**Comment by justinmk** [comment]
@brcolow Please remove NA from [778](https://github.com/vim/vim/commit/ae2fe73abc954b8fd3dbd7994daa8e31e6690bd0), it is needed once we merge 7.4.754.

**Comment by brcolow** [comment]
@justinmk Done.
@wsdjeg Happy new year to you as well.

**Comment by justinmk** [comment]
> Question on 989: Using xfree instead of vim_free as done in the original patch is correct? 

xfree is correct, vim_free was replaced by xfree for us.

> 982 patch NA question: 
> believe they are only specified in exactly one location, and this is only for legacy tests that have not yet been ported). Please advise :).

:+1:  Right.

> 964 patch NA question:
> Its related to test 87, which as per @justinmk 's comments above was deleted, so this can also be marked as NA?

:+1: Yes. Also the whole "shadow dir" thing probably isn't relevant to us.

> (P.S. sorry for some of the renaming noise, etc...I'm quite new to this)

A concise, scoped name like "version.c: update" is preferable.

**Comment by brcolow** [comment]
Updated to reflect the two additional patches (964 and 982) as NA.

**Comment by justinmk** [comment]
For the commit message too :)

**Comment by justinmk** [comment]
Oh, I see you're porting actual patches here too. That's fine, but in that case the minimal commit message should be like this:

```
vim patches

vim-patch:7.4.955
vim-patch:7.4.974
vim-patch:7.4.975
vim-patch:7.4.989
```

This allows us to grep the logs easily to see if/when a patch was applied.

Also in the future please separate actual patches in separate commits. Explained here: https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim

**Comment by brcolow** [comment]
Okay, thanks, fixed. I will certainly adhere to that in the future - thank you for not making me try and struggle through separating this PR into separate commits, as that would surely lead to chaos given my limited abilities.

**Comment by Shougo** [comment]
Nice work!
@brcolow I suggest for you to update it to 7.4.1050.
I have analysed the code.

New NA patches list until 7.4.1050:

1002 Windows Test
1003 Travis CI
1004 Makefile
1008 Drop OS/2 support
1009 ARCHIE is already removed in neovim
1011 if_perl
1012 if_python
1014 Cygwin support
1016 Drop OS/2 support
1019 Makefile fix
1020 Windows fix
1021 Makefile fix
1022 Remove outdated description
1023 Makefile fix
1024 Windows fix
1025 Makefile fix
1026 Makefile fix
1028 Nsis version file
1031 if_python
1033 Windows fix
1038 configure fix
1040 tee.c
1043 INSTALL_mac.txt
1044 ifdef FEAT_EVAL
1045 Travis fix

**Comment by brcolow** [comment]
@Shougo Done. Thank you very much. Would you like me to add your name to the commit message in some way?

**Comment by Shougo** [comment]
Yes, thank you!

**Comment by Shougo** [comment]
Thank you!

**Comment by justinmk** [comment]
QB failure is luajit download error.

File: src/nvim/diff.c [hunk 2]
```diff
@@ -1020,8 +1021,18 @@ void ex_diffsplit(exarg_T *eap)
...
     if (curwin != old_curwin) {
       // Set 'diff', 'scrollbind' on and 'wrap' off.
-      diff_win_options(curwin, TRUE);
-      diff_win_options(old_curwin, TRUE);
+      diff_win_options(curwin, true);
+      if (win_valid(old_curwin)) {
+        diff_win_options(old_curwin, true);
+
+        if (buf_valid(old_curbuf)) {
+            curwin->w_cursor.lnum = diff_get_corresponding_line(
```
Comment (justinmk) [inline src/nvim/diff.c]: comment missing from original patch. I'll add it when merging.

File: src/nvim/ex_cmds.c [hunk 3]
```diff
@@ -495,10 +495,15 @@ void ex_sort(exarg_T *eap)
...
         || (sort_ic ? STRICMP(s, sortbuf1) : STRCMP(s, sortbuf1)) != 0) {
-      if (ml_append(lnum++, s, (colnr_T)0, FALSE) == FAIL)
+      // Copy the line into a buffer, it may become invalid in
+      // ml_append(). And it's needed for "unique".
+      STRCPY(sortbuf1, s);
+      if (ml_append(lnum++, sortbuf1, (colnr_T)0, false) == FAIL) {
         break;
-      if (unique)
+      }
+      if (unique) {
```
Comment (justinmk) [inline src/nvim/ex_cmds.c]: This was deleted in the original patch, I'm guessing git rebase issue?

File: src/nvim/version.c [hunk 4]
```diff
@@ -82,32 +82,32 @@ static int included_patches[] = {
...
   // 977,
   // 976 NA
-  // 975,
+  975,
   // 974,
   // 973,
-  // 972,
-  // 971,
-  // 970,
+  972,
```
Comment (justinmk) [inline src/nvim/version.c]: This patch is missing from this PR

**Comment by justinmk** [comment]
@brcolow @Shougo The NA patches suggested in https://github.com/neovim/neovim/pull/3919#issuecomment-169319129 are not present in this PR. But I don't want to have to re-review it, so I merged it as-is (with the changes in my comments above).

**Comment by Shougo** [comment]
@brcolow @justinmk 
OK.
Should I update version.c to the latest version?

**Comment by justinmk** [comment]
:+1:

**Comment by Shougo** [comment]
I did it.

**Comment by watiko** [comment]
Sorry, this reference log is wrong.

**Pull request closed without merging**